FEAT: Add support for Postgres UDFs

scottcode opened the pull request:
The following functionality is included: 
* ability to use user-defined functions (UDFs) 
* ability to use UDFs that are already defined in the database (I just want to refer to them and use them)
* ability to wrap a python function, automatically define it as a PL/Python function in the database, and be able to use it with ibis objects
* changed Docker image to support testing of Postgres PL/Python UDF functionality in the ibis test suite
* Tests for UDF functionality

The current API looks like this. Comments and feedback are welcome. 

```
import ibis.expr.datatypes as dt
from ibis.sql.postgres.udf.api import existing_udf, func_to_udf

my_udf = existing_udf(
    'my_udf',
    input_types=[dt.String()],
    output_type=dt.Integer(),
    schema='my_schema'
)

expr = table[my_udf(table['str_col']).name('int_col_output')]

def mult_a_b(a, b):
    return a * b

mult_a_b_udf = func_to_udf(
    sqlalchemy_engine,
    mult_a_b,
    (dt.Int32(), dt.Int32()),
    dt.Int32(),
    schema='my_schema',
    overwrite=True
)

expr2 = table[mult_a_b_udf(table['int_col1'], table['int_col2']).name('int_col_output')]
```

scottcode commented:
It appears that some of the failures were due to...

* timeouts from `impala` startup (in LinuxBuildDocs pipeline). 
* All of the Windows pipelines failed due to some problem creating the PL/Python extension: `could not access file "$libdir/plpython2"`. I wonder if there's some Windows-specific problem not resolving an environment variable due to the `$` notation instead of `%VAR%`
* Conda build (I don't understand enough about feedstock.py and conda builds to know what's going on. I don't think anything in my PR itself would have affected this process.

scottcode commented:
Figured out why it's probably failing on the conda build. The conda build doesn't spin up the backends and is supposed to exclude tests that rely on them. It excludes things marked with `pytest.mark.postgresql`, but I hadn't marked my `test_udf.py` module with that. My new commit adds that.

scottcode commented:
Aha, I figured out that the Windows tests use a different installation of Postgres, one which does not include any extensions (neither PostGIS nor pl/python). The Windows pipeline was already excluding postgis tests. I tweaked my tests to not try enabling PL/Python by default, and I added `postgis` marker so the `plpython` tests would be excluded along with the postgis ones. Eventually we need to update the Azure Windows pipeline to exclude a new marker called `postgres_extensions`, as this would cover both postgis and pl/python. 

I included the updated `windows.yml` in this commit. @cpcloud, what is necessary to update the Azure pipelines?

I am eager to hear feedback from the community.

cpcloud commented:
@scottcode Wow, this is really interesting, thanks for doing this. I'm going to review a bit today, and then the rest over the weekend.

cpcloud commented on ibis/sql/postgres/tests/test_udf.py:
```diff
@@ -0,0 +1,173 @@
+    )
+    con.con.execute(sql_define_udf.format(schema=test_schema))
+    con.con.execute(sql_define_py_udf.format(schema=test_schema))
+    yield con
```
This needs to be:

```python
try:
    yield con
finally:
    <teardown code>
```

The way you've written it the teardown code will not execute if the function yielded to raises an exception.

scottcode replied:
That makes sense. I'll change that.

cpcloud commented on ibis/sql/postgres/tests/test_udf.py:
```diff
@@ -0,0 +1,173 @@
+        schema_exists = df_result['n_recs'].iloc[0] > 0
+        tries += 1
+    assert tries > 0
+    global test_schema
```
Why did you decide to make this a global variable as opposed to a fixture or a function that returns the schema?

scottcode replied:
Do fixtures re-execute every time they are used, or are they cached? Because I don't want it to try re-creating the test_schema every time the fixture is used.

cpcloud replied:
You can specify the scope at which they run, the "highest" scope being `'session'` which runs exactly once per test session. You can specify this using `@pytest.fixture(scope='session')`.

We have a number of tests that perform stateful manipulations (like creating a table) on databases (Impala in particular) and we're able to run those in parallel, so in theory we should be able to do the same with postgres.

scottcode replied:
Fixed. No longer using a global and doing it as a fixture with session scope

cpcloud commented on ibis/sql/postgres/tests/test_udf.py:
```diff
@@ -0,0 +1,173 @@
+return len(x)
+$$;"""
+
+sql_define_udf = """CREATE OR REPLACE FUNCTION {schema}.custom_len(x varchar)
```
Make all of these module globals into `pytest.fixture`s, it's easier to see what's being used and where it's being used.

cpcloud commented on ibis/sql/postgres/tests/test_udf.py:
```diff
@@ -0,0 +1,173 @@
+def test_sql_length_udf_worked(table):
+    """Test creating ibis UDF object based on existing UDF in the database"""
+    # Create ibis UDF objects referring to UDFs already created in the database
+    custom_length_udf = ibis.sql.postgres.udf.api.existing_udf(
```
If `existing_udf` is a user-facing API then it should be exposed in `ibis.sql.postgres.api` and you should refer to it as `ibis.postgres.existing_df`.

cpcloud commented on ibis/sql/postgres/tests/test_udf.py:
```diff
@@ -0,0 +1,173 @@
+    # Create ibis UDF objects referring to UDFs already created in the database
+    custom_length_udf = ibis.sql.postgres.udf.api.existing_udf(
+        'custom_len',
+        input_types=[ibis.expr.datatypes.String()],
```
Use `dt.string` and `import ibis.expr.datatypes as dt`.

cpcloud commented on ibis/sql/postgres/tests/test_udf.py:
```diff
@@ -0,0 +1,173 @@
+    custom_length_udf = ibis.sql.postgres.udf.api.existing_udf(
+        'custom_len',
+        input_types=[ibis.expr.datatypes.String()],
+        output_type=ibis.expr.datatypes.Integer(),
```
`Integer` is an abstract class that is the base of all other integer types. This should be something like `dt.int64`, `dt.int32`, etc.

cpcloud commented on ibis/sql/postgres/tests/test_udf.py:
```diff
@@ -0,0 +1,173 @@
+    mult_a_b_udf = ibis.sql.postgres.udf.api.func_to_udf(
+        con_for_udf.con,
+        mult_a_b,
+        (datatypes.Int32(), datatypes.Int32()),
```
use `dt.int32` here.

cpcloud commented on ibis/sql/postgres/tests/test_udf.py:
```diff
@@ -0,0 +1,173 @@
+        con_for_udf.con,
+        mult_a_b,
+        (datatypes.Int32(), datatypes.Int32()),
+        datatypes.Int32(),
```
`dt.int32`

cpcloud commented on ibis/sql/postgres/tests/test_udf.py:
```diff
@@ -0,0 +1,173 @@
+def test_func_to_udf_smoke(con_for_udf, table):
+    """Test creating a UDF in database based on Python function
+    and then creating an ibis UDF object based on that"""
+    mult_a_b_udf = ibis.sql.postgres.udf.api.func_to_udf(
```
I'm not a huge fan of this API. What do you think about this?

```python
@ibis_client.udf([dt.int32, dt.int32], dt.int32, schema='my_schema', overwrite=True)
def mult_a_b(a, b):
    return a * b
```

scottcode replied:
I'm totally open to changing the API. I was so focused on just getting something that worked, and was actually hoping you and the community would chime in on how to improve the API. I'll work on refactoring. 

What would you prefer the API be for existing UDFs? Still have a function like the `existing_udf` that I have? Would you prefer a different name for it?

scottcode replied:
My most recent push [47aacdd7](https://github.com/scottcode/ibis/commit/47aacdd756a70eb9d81374b73ff4e212ce0d8d01) adds your preferred `@ibis_client.udf(...)` decorator API.

cpcloud commented on ci/load-data.sh:
```diff
@@ -4,7 +4,7 @@ CWD="$(dirname "${0}")"
 declare -A argcommands=([sqlite]=sqlite
                         [parquet]="parquet -i"
-                        [postgres]=postgres
+                        [postgres]='postgres --plpython'
```
If you set `default=True` then you don't need to change this file.

cpcloud commented on ibis/sql/postgres/udf/api.py:
```diff
@@ -0,0 +1,218 @@
+    (as is used in SQL defining UDF signatures)
+
+    :param type_: str, a Python data type,
+                  or an ibis DataType (instance or class)
```
This is handling too many input types. This function should accept an ibis datatype, and any validation should be done in the caller.

scottcode replied:
Yeah, I agree that accepting too many input types is not great. I was trying to mimic the multiple dispatch in `ibis.bigquery.datatypes` without resorting to the same `multipledispatch` approach used there. I'll simplify this to just take ibis datatypes.

scottcode replied:
Multiple dispatch no longer applicable, because I'm now only allowing ibis DataTypes as parameters, and I changed the way I'm translating to the SQL type strings.

cpcloud commented on ibis/sql/postgres/udf/api.py:
```diff
@@ -0,0 +1,218 @@
+                 input_types,
+                 output_type,
+                 schema=None,
+                 parameter_names=None):
```
Let's change `parameter_names` to `parameters`.

cpcloud commented on ibis/sql/postgres/udf/api.py:
```diff
@@ -0,0 +1,218 @@
+                 output_type,
+                 schema=None,
+                 parameter_names=None):
+    """Create a ibis function that refers to an existing Postgres UDF already
```
`a` -> `an`

cpcloud commented on ibis/sql/postgres/udf/api.py:
```diff
@@ -0,0 +1,218 @@
+    if parameter_names is None:
+        parameter_names = ['v{}'.format(i) for i in range(len(input_types))]
+    else:
+        assert len(input_types) == len(parameter_names)
```
This should raise some kind of exception other than `AssertionError`.

scottcode replied:
Changed it to raise a `ValueError`

cpcloud commented on ibis/sql/postgres/udf/api.py:
```diff
@@ -0,0 +1,218 @@
+    udf_node = create_udf_node(name, udf_node_fields)
+
+    def _translate_udf(t, expr):
+        func_obj = sqlalchemy.func
```
Use `sa` and `import sqlalchemy as sa` at the top of the module.

scottcode replied:
done

cpcloud commented on ibis/sql/postgres/udf/api.py:
```diff
@@ -0,0 +1,218 @@
+
+    def _translate_udf(t, expr):
+        func_obj = sqlalchemy.func
+        if schema is not None:
```
Can we set `schema` to some kind of valid default (I think it's actually `'default'` in postgres) so that we don't need to first check if `schema` is None? You could then write this as `func_obj = getattr(getattr(sa.func, schema), name)`.

scottcode replied:
I'd like to keep the possibility open for not qualifying the schema. While there usually is a default schema called `public` in Postgres, there is also a way (in postgres and in `sqlalchemy`) of setting the schema search path so that another non-`public` schema is used when unqualified.

cpcloud commented on ibis/sql/postgres/udf/api.py:
```diff
@@ -0,0 +1,218 @@
+        for name, type_ in zip(parameter_names, in_types)
+    )
+    return_type = get_sqltype(out_type)
+    func_definition = inspect.getsource(python_func)
```
Something to be aware of is that `inspect.getsource` only works if `python_func` is defined in a file. This means that in a vanilla `python` session (*not* IPython) you won't be able to define a UDF using this mechanism.

This isn't a showstopper IMO; the utility of this feature is high enough that it's not a hard requirement to support every possible scenario that folks may use it in.

scottcode replied:
I played around with a scrap script/module, and `inspect.getsource` **_is_** still able to get the source, even when in vanilla python and not in IPython. 

That said, there are certainly still edge cases that my current implementation doesn't cover. For instance, it won't work if the function uses any variables defined outside the scope of the function itself or any modules imported outside the function. There could be ways of automatically detecting imports that need to be inserted or sucking in globals, but I would propose that **for now it can be the user's responsibility to make sure their function is entirely self-contained**.

cpcloud commented on ibis/sql/postgres/udf/api.py:
```diff
@@ -0,0 +1,218 @@
+# type mapping based on: https://www.postgresql.org/docs/10/plpython-data.html
+sql_default_type = 'VARCHAR'
+
+pytype_sql = {
```
I think we'd probably want to use `sqlalchemy.types` here, to depend on the 10+ years of work spent making sure that its types compile to the correct postgres type.

scottcode replied:
I switched to using `ibis` and `sqlalchemy`s builtin mechanisms for translating types.

scottcode commented:
Hmm, this is frustrating. I fixed the Windows issues, rebased on Master, and now the linux stuff fails, but only for geospatial-related stuff (not my new UDF code). I'll try to figure this out, but I'm certainly confused at the moment.

cpcloud commented on ibis/sql/postgres/udf/api.py:
```diff
@@ -0,0 +1,218 @@
+    signature = inspect.signature(python_func)
+    parameter_names = signature.parameters.keys()
+    if in_types is None:
+        raise NotImplementedError('inferring in_types not implemented')
```
Curious what it means to infer the input types here?

scottcode replied:
Now that Python 3 supports type annotations in function signatures, my plan is to add logic that infer the ibis types from the Python types in the annotations. See example below for getting the Python types from a signature. 

```
import inspect

def total(price: float, quantity: int) -> float:
    return price * quantity

sig = inspect.signature(total)

price_pytype = sig.parameters['price'].annotation
return_pytype = sig.return_annotation
```

If any ibis types are explicitly passed in, those would take precedence. But if none are passed in yet the function has type annotations, it could map those python types to ibis types and use the inferred ones. 

If they were annotated with ibis types already we could skip inference. What I mean by inference is that there's already a [mapping between Python and Postgres types](https://www.postgresql.org/docs/10/plpython-data.html) defined in the PL/Python extension itself, so we can simply use that specification to do the inference.

scottcode replied:
Perhaps I should just remove the placeholders for inferring types since it's not yet implemented anyway. That could be a subsequent issue and pull request, and it would keep the code looking cleaner in the meantime.

scottcode replied:
I just removed the placeholders.

cpcloud commented on ibis/sql/postgres/udf/api.py:
```diff
@@ -0,0 +1,218 @@
+    if in_types is None:
+        raise NotImplementedError('inferring in_types not implemented')
+    if out_type is None:
+        raise NotImplementedError('inferring out_type not implemented')
```
Similar question to the above: what does it mean to infer the output type?

cpcloud commented:
@scottcode Don't spend any time on fixing the geospatial stuff, I'm about to merge https://github.com/ibis-project/ibis/pull/1872 which should fix it.

scottcode commented on ci/docker-compose.yml:
```diff
@@ -2,12 +2,12 @@ version: '3'
 
   postgres:
-    image: mdillon/postgis
+    image: shajekpivotal/postgres-9.5-postgis-2.5
```
That's a great idea. I just created a repo I could control in order to make progress.

scottcode replied:
If you want to create that other repo within `ibis-project/docker-postgres` and give me access to it an `ibis-project` Dockerhub account, I'd be happy to work on transitioning the new postgres image.

scottcode replied:
I just created a separate repo that streamlines the Postgres Dockerfile into a single code repository. That code repository and the corresponding Dockerhub image can easily migrated to an `ibis-project` account whenever that is made available. 

For reference, the code is in [github.com/autoscott/ibis-docker-postgres](https://github.com/autoscott/ibis-docker-postgres), and the Docker image is currently hosted in [hub.docker.com/r/shajekpivotal/ibis-docker-postgres-9.5](https://hub.docker.com/r/shajekpivotal/ibis-docker-postgres-9.5). 

I just updated updated the `docker-compose.yml` to use this new image.

scottcode commented on ibis/sql/postgres/tests/test_udf.py:
```diff
@@ -0,0 +1,173 @@
+    test schema name"""
+    schema_name = '{}_{}'.format(
+        basename,
+        ''.join([random.choice(string.ascii_lowercase) for i in range(6)])
```
Because the testing apparatus often runs in parallel mode, which means parallel processes that don't share globals. No shared globals, no way to do unique incrementing.... (unless we do a Postgres sequence object and base it on the database). Without adding the random strings and checking for existence, when the tests were run in parallel then I was getting contention issues.

cpcloud replied:
You can decorate a fixture with `@pytest.fixture(scope='session')` to make the fixture execute exactly once per session. We are mutating state in other backends (notably Impala) so in theory we should be able to do this with postgres as well.

cpcloud replied:
This still needs to be addressed. Can you try using `scope='session'`?

scottcode replied:
Just pushed changes with the session-based scope. Also took out the random character-generation method of keeping test schema separate. Commit [e1ee3bbe60](https://github.com/scottcode/ibis/commit/e1ee3bbe601d3d715f6c6a352543c4594e2fa4a1)

scottcode commented on ibis/sql/postgres/tests/test_udf.py:
```diff
@@ -0,0 +1,173 @@
+    testing"""
+    schema_exists = True
+    tries = 0
+    while schema_exists and tries < max_tries:
```
I just wanted to take into account the possibility that multiple parallel test instances may have created the same random string. It is unlikely (1 in 26**6) but possible. Max tries is probably not necessary, but I was nervous about creating a `while` loop without an explicit cap in case something isn't working right.

scottcode replied:
And as I mentioned in a separate comment, when running tests in parallel mode there can be contention for the same name, and I wasn't sure of a way to ensure a unique sequence ID across processes.

scottcode replied:
Just pushed changes with the session-based scope that took out the multiple attempts at creating a unique schema. Commit [e1ee3bbe60](https://github.com/scottcode/ibis/commit/e1ee3bbe601d3d715f6c6a352543c4594e2fa4a1)

scottcode commented on ci/datamgr.py:
```diff
@@ -228,7 +228,13 @@ def parquet(tables, data_directory, ignore_missing_dependency, **params):
+@click.option(
+    '--plpython/--no-plpython',
+    help='Create PL/Python extension in database',
+    is_flag=True,
```
The only reason I defaulted to False is that I wasn't sure I could change the Azure Windows pipeline. I can change default to `True` and would just need to tweak the invocation of `load_data` in the Azure pipeline.

scottcode commented:
Thanks @cpcloud for all the great feedback! I just wanted to call out separately how much I appreciate it. I'm starting to make my way through all of it.

scottcode commented:
@cpcloud I've addressed most of your comments and pushed a new version. The `@ibis_client.udf` decorator is still in progress, and there might be a few other outstanding issues, but wanted you to be able to see all the progress.

cpcloud commented on ibis/sql/postgres/udf/api.py:
```diff
@@ -0,0 +1,309 @@
+        ast.NodeVisitor.generic_visit(self, node)
+
+
+def remove_decorators(funcdef_source):
```
It's a bit concerning that we have to effectively delete code to make this work. The decorator API is nice, but I think having to use the `ast` module to manipulate the source is too much magic.

cpcloud replied:
We could leave the implementation as is, remove the decorator deletion code and document that use of the `udf` method as decorator will not work because of `$REASONS`.

scottcode replied:
I can do that (remove decorator deletion code and add some documentation for `$REASONS`). I could also add a check that if it has decorators to throw an exception or warning.

scottcode replied:
I see two patterns possible here. We could choose not to support decorator syntax but still have a two-step process to conform with the other UDF APIs:

```python
my_udf = ibis_client.udf([dt.int32], dt.int32)(my_function)
```

Or include the function object in the first call:

```python
my_udf = ibis_client.udf(my_function, [dt.int32], dt.int32)
```

I prefer the latter streamlined version. @cpcloud, would you be okay with that?

scottcode replied:
I just pushed a new version that takes out the deletion and implements the latter non-decorator form (`ibis_client.udf(func, [dt.int32], dt.int32)`)

scottcode commented on ibis/sql/postgres/udf/api.py:
```diff
@@ -0,0 +1,309 @@
+    )
+
+
+class UdfDecorator(object):
```
Just switched it to your suggestion of using nested function and closure. _The class-based decorator was a holdover from my feverish attempt at figuring out how to get `inspect.getsource` to exclude decorators, but it didn't help anyway._  😜

scottcode commented on ibis/sql/postgres/client.py:
```diff
@@ -200,3 +201,33 @@ def list_tables(self, like=None, database=None, schema=None):
+        schema : str
+            optionally specify the schema in which to define the UDF
+        overwrite : bool
+            replace UDF in database if already exists
```
Agreed. The rename has been pushed

scottcode commented on ibis/sql/postgres/client.py:
```diff
@@ -200,3 +201,33 @@ def list_tables(self, like=None, database=None, schema=None):
+            an ibis function
+        """
+
+        return UdfDecorator(
```
No longer applicable. Got rid of class-based decorator anyway, as per one of your other comments.

scottcode commented on ibis/sql/postgres/udf/__init__.py:
```diff
@@ -0,0 +1,3 @@
+from ibis.sql.postgres.udf.api import existing_udf, func_to_udf, UdfDecorator
+
+__all__ = ('existing_udf', 'func_to_udf', 'UdfDecorator')
```
I guess not, if we are going to keep the `client.udf()` method (even if we don't keep the decorator functionality).

scottcode replied:
I removed them from the `__init__.py`s

scottcode commented on ibis/sql/postgres/udf/api.py:
```diff
@@ -0,0 +1,309 @@
+
+    def generic_visit(self, node):
+        if hasattr(node, 'lineno'):
+            self.non_decorator_lines.append(node.lineno)
```
Changed to only keep track of the min line number

scottcode commented on ibis/sql/postgres/udf/api.py:
```diff
@@ -0,0 +1,309 @@
+
+    def visit_FunctionDef(self, node):
+        self.decorator_lines.extend(
+            [n.lineno for n in node.decorator_list]
```
Removed

scottcode commented on ibis/sql/postgres/udf/api.py:
```diff
@@ -0,0 +1,309 @@
+        self.decorator_lines.extend(
+            [n.lineno for n in node.decorator_list]
+        )
+        for field1, node1 in ast.iter_fields(node):
```
Changed to `func_field` and `func_node`

scottcode commented:
@cpcloud  Looks like there's some problem with the Azure pipeline setup. Looking at the logs it says 

> /azure-pipelines.yml: Could not retrieve file content for /ci/azure/linux.yml from repository self hosted on https://github.com/ using commit e97f541e7c4d0eff7422a75584d59a4dbc6f9cfe. GitHub reported the error, "Bad credentials"

scottcode commented:
@cpcloud All the CI tests passed, and I believe I've addressed all the changes you've requested so far. Are there any others you'd like me to make?

cpcloud commented on ibis/sql/postgres/tests/test_udf.py:
```diff
@@ -0,0 +1,225 @@
+            [dt.int32, dt.int32],
+            dt.int32,
+            schema=test_schema,
+            replace=True)
```
Formatting.

scottcode replied:
Just changed via `black`

cpcloud commented on ibis/sql/postgres/udf/api.py:
```diff
@@ -0,0 +1,218 @@
+                out_type,
+                schema=None,
+                replace=False,
+                name=None):
```
Formatting.

cpcloud commented on ibis/sql/postgres/udf/api.py:
```diff
@@ -0,0 +1,218 @@
+    )
+    if func_definition.strip().startswith('@'):
+        raise PostgresUDFError(
+            'Use of decorators on function to be turned into Postgres UDF '
```
`on function` -> `on a function`.

This error message is a bit verbose. Can you shave it down a bit?

scottcode replied:
I moved the finer detail into a comment in the code right after the error message.

cpcloud commented on ibis/sql/postgres/udf/api.py:
```diff
@@ -0,0 +1,218 @@
+    return_type = ibis_to_postgres_str(out_type)
+    # If function definition is indented extra,
+    # Postgres UDF will fail with indentation error.
+    # Also, need to remove decorators, because they
```
You're no longer removing decorators, right?

scottcode replied:
Correct. I just deleted the comment referring to removing decorators.

cpcloud commented on ibis/sql/postgres/udf/api.py:
```diff
@@ -0,0 +1,218 @@
+        )
+        for name, type_ in zip(parameter_names, in_types)
+    )
+    return_type = ibis_to_postgres_str(out_type)
```
We should ultimately make a SQLAlchemy construct and go through _its_ compiler rather than stringifying SQLAlchemy types and formatting them in here. Eventually that might actually be a nice upstream contribution as well.

cpcloud replied:
Let's make a follow up issue for that.

scottcode replied:
Great suggestion! I created #1890 for this.

cpcloud commented on ibis/sql/postgres/udf/api.py:
```diff
@@ -0,0 +1,218 @@
+
+    Returns
+    -------
+    wrapper : Callable
```
This can just be `Callable`. There's really no point in having a named variable in the `Returns` documentation since it's subject to change and irrelevant to users.

cpcloud commented on ibis/sql/postgres/udf/api.py:
```diff
@@ -0,0 +1,218 @@
+
+    Returns
+    -------
+    result : type
```
Remove the `result`.

scottcode commented on ibis/sql/postgres/tests/test_udf.py:
```diff
@@ -0,0 +1,225 @@
+    return a * b
+
+
+def test_func_to_udf_smoke(con_for_udf, test_schema, table):
```
Took off `_smoke`. (I originally called it a smoke test because it tests end-to-end functionality including going through the backend, but most of the tests are doing that, so you're right, it doesn't make sense to call it a smoke test.)

scottcode commented on ibis/sql/postgres/tests/test_udf.py:
```diff
@@ -0,0 +1,225 @@
+# Tests
+
+
+def test_sql_length_udf_worked(test_schema, table):
```
Renamed to `test_existing_sql_udf`

scottcode commented on ibis/sql/postgres/client.py:
```diff
@@ -200,3 +201,41 @@ def list_tables(self, like=None, database=None, schema=None):
+            out_type,
+            schema=None,
+            replace=False,
+            name=None):
```
I previously had trouble installing them so wasn't using them, but I've figured it out now and have run `black` on the modules I've been touching.

scottcode commented:
The only errors at this point are in the Lint step (only in Windows Python 3.6 and 3.7). It is unclear to me why it is erroring out; the logs don't give any hints. Running the lint locally everything checks out.

scottcode commented:
Only error at this point is backend waiter timeout on LInux Python 3.5 after 10 minutes.

cpcloud commented:
Yeah I just kicked it off again to see if it was a transient error.

scottcode commented:
Looking at logs from several builds that timed out at that step, it's always only Impala that it's still waiting on. I wonder if there's a way to speed up Impala's load time.

scottcode commented:
It looks like the current image for Impala includes unnecessarily installing and starting up Kudu (since in our `docker-compose` we start them up from separate images/containers and use those anyway). Maybe we could cut the following out of the Impala startup to speed it up. 

[ibis-project/docker-impala/etc/supervisord.conf](https://github.com/ibis-project/docker-impala/blob/master/etc/supervisord.conf)

```
[program:kudu-master]
command=/bin/bash -c 'exec /etc/init.d/kudu-master start'
autostart=false
startsecs=0

[program:kudu-tserver]
command=/bin/bash -c 'exec /etc/init.d/kudu-tserver start'
autostart=false
startsecs=0

```

scottcode commented:
Think I've figured out why this still sometimes fails while waiting for backends. Just pushed my fix. Fingers crossed that it passes the Azure pipelines. Below is my commit message explanation. 

> **Respect dependency of impala on postgres service**
> 
> The impala service actually relies on there being an instance of
> postgres available to use as its metadata store. It has a timer to
> check up to 120 seconds and then impala startup hangs indefinitely or
> fails.
> 
> When we switched to a new image for postgres that includes PL/Python,
> this pushed the startup time just long enough to cause impala startup
> to sometimes fail (unpredictably).
> 
> Instead of increasing the wait time hard-coded in the Impala image,
> it seems better/easier to enforce the dependency with the
> `docker-compose.yml`.
> 
> See https://github.com/ibis-project/docker-impala/blob/master/bin/start-impala.sh line 21 at revision dcb7388
> 
> `/wait-for-it.sh postgres:5432 -t 120`

scottcode commented:
@cpcloud I got it all green 😃

cpcloud commented on ibis/sql/postgres/udf/api.py:
```diff
@@ -0,0 +1,213 @@
+    return wrapped
+
+
+def func_to_udf(
```
If this is going to be a public API, can we call it `udf` and have it take an ibis client instead of `conn`? That way `client.udf(...) == `udf(client, ...)` and folks can choose whether they want a functional or object-oriented interface.

scottcode replied:
That's a good idea. I just made this change and re-pushed.

scottcode commented:
Phillip, I know you already saw this, but for others' reference the issue for the Postgres Docker image is #1853.

cpcloud commented:
Kaboom. Merge time.

cpcloud commented:
Thanks @scottcode, keep 'em coming!
